File: 13_candy.cpp
class Solution {
public:
    int candy(vector<int>& ratings) {
        int n = (int)ratings.size();
        vector<pair<int,int>>A;
        for(int i=0;i<n;i++){
            A.push_back({ratings[i],i});
        }
        sort(A.begin(),A.end());
        vector<int> ans(n,0);
        for(auto i : A){
            int idx=i.second;
            int mx=0;
            
            if(idx>0 && ratings[idx]>ratings[idx-1]){
                mx=max(ans[idx-1],mx);
            }
            if(idx<n-1 && ratings[idx]>ratings[idx+1]){
                mx=max(ans[idx+1],mx);
            }
            ans[idx]=mx+1;
            // cout<<"mx "<<ans[idx]<<" i.se " <<i.second<<endl;
        }
        // for(int i=0;i<n;i++)cout<<ans[i]<<" ";
        // cout<<endl;
        int res = accumulate(ans.begin(),ans.end(),0);
        return res;
    }
};

/*

this is a lengthier implementation but more intuitive

we allocate the cady to the lowest height guy and then some more to the guy with higher height
and so on


*/

File: 13_candy.go
func candy(A []int) int {
    n := len(A)
    left, right := make([]int, n), make([]int, n)
    var ans int
    for i:=0;i<n;i++{
        left[i],right[i]=1,1
    }
    for i:=1;i<n;i++{
        if A[i] > A[i-1]{
            left[i]=left[i-1]+1
        }
    }
    for i:=n-2;i>=0;i--{
        if A[i]>A[i+1]{
            right[i]=right[i+1]+1
        }
    }
    max := func(a, b int) int {
        if a > b { return a }
        return b
    }

    for i:=0;i<n;i++{
        ans += max(left[i],right[i])
    }
    return ans
}

/*


fixing a left value fixes everything from left to right
and similarly, fixing a right value fixes everything from right to left

*/
